webui/react/src/pages/Cluster.tsx

```diff
@@ -0,0 +1,55 @@
+import React, { useCallback } from 'react';
+
+import Grid from 'components/Grid';
+import Page from 'components/Page';
+import ResourceChart from 'components/ResourceChart';
+import Spinner from 'components/Spinner';
+import AgentsCtx from 'contexts/Agents';
+import emptyMessage from 'styles/emptyMessage.module.scss';
+import { Resource } from 'types';
+import { categorize } from 'utils/data';
+
+import css from './Cluster.module.scss';
+
+const Cluster: React.FC = () => {
+  const agents = AgentsCtx.useStateContext();
+
+  const getResources = useCallback(() => {
+    if (!agents.data) return {};
+    const resourceList = agents.data?.map(a => a.resources).reduce((a, c) => {
+      a.push(...c); return a;
+    });
+    return categorize(resourceList, (res: Resource) => res.type);
+  }, [ agents ]);
+
+  const resourceDict = getResources();
+
+  const availableResTypes = Object.keys(resourceDict);
+
+  let unhappyView: React.ReactNode;
+  if (!agents.data) {
+    unhappyView = <Spinner />;
+  } else if (agents.data && agents.data.length === 0) {
+    unhappyView = (<div className={emptyMessage.base}>
+      No agents connected.
+    </div>);
+  } else if (availableResTypes.length === 0) {
+    unhappyView = (<div className={emptyMessage.base}>
+      No slots available.
+    </div>);
+  }
```

Comment:
suggestion: putting `<Grid>` into the same conditional render logic
```
let contentView = <Spinner />;
if (agents.data && agents.data.length === 0) {
  ... <EmptyMessage />
} else if (availableResourceTypes.length ===0 {
  ... <EmptyMessage />
} else if (availableResourceTypes.length > 0) {
  ... <Grid stuff />
}
```

Reply:
would you assign them to a variable and then render that? 
I did some changes here. could you take another look?